### PR TITLE
Invalidate ExecCtx now before computing timeouts in all repeating timer events using a WorkSerializer or combiner

### PR DESCRIPTION
**Pull request by apolcyn**
This is a tactical alternative to https://github.com/grpc/grpc/pull/26064

Related: https://github.com/grpc/grpc/issues/26079

I'd still like to find a structural change to fix https://github.com/grpc/grpc/issues/26079 without needing to pay close attention to whether or not that kind of issue can affect a certain code path, but this change should fix the user issue mentioned in b/185205431 in the meantime.

### CODE REVIEW

File: src/core/ext/filters/client_channel/backup_poller.cc
Comment (markdroth) [inline src/core/ext/filters/client_channel/backup_poller.cc:134]: This doesn't enter a `WorkSerializer`.

File: src/core/ext/filters/client_channel/health/health_check_client.cc
Comment (markdroth) [inline src/core/ext/filters/client_channel/health/health_check_client.cc:141]: This doesn't enter a `WorkSerializer`.

File: src/core/ext/filters/client_channel/lb_policy/priority/priority.cc
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/priority/priority.cc:639]: This is not a repeated timer, so there can't be a loop here.

File: src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc:562]: This is not a repeated timer, so there can't be a loop here.

File: src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_manager.cc
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_manager.cc:508]: This is not a repeated timer, so there can't be a loop here.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc
Comment (markdroth) [inline src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc:491]: This is not a repeated timer, so there can't be a loop here.

File: src/core/ext/filters/client_channel/subchannel.cc
Comment (markdroth) [inline src/core/ext/filters/client_channel/subchannel.cc:953]: This doesn't enter a `WorkSerializer`.

File: src/core/ext/filters/client_channel/retry_filter.cc
Comment (markdroth) [inline src/core/ext/filters/client_channel/retry_filter.cc:2132]: This doesn't enter a `WorkSerializer`.

File: src/core/ext/filters/fault_injection/fault_injection_filter.cc
Comment (markdroth) [inline src/core/ext/filters/fault_injection/fault_injection_filter.cc:431]: This doesn't enter a `WorkSerializer`.

File: src/core/ext/transport/chttp2/transport/writing.cc
Comment (markdroth) [inline src/core/ext/transport/chttp2/transport/writing.cc:78]: I'm not sure about this one -- @yashykt should review this one.

**Review by markdroth — COMMENTED** [review]
My understanding from the bug is that in order to trigger this problem, we need to have both (a) a repeating timer where the timer callback reschedules the timer and (b) the timer callback entering the `WorkSerializer`.  Many of the places where you're changing the code are missing one of those two criteria.

I think we need to discuss the right way to solve the problem more generally.  Until then, let's add this work-around only in the cases where the specific problem we're seeing is occurring.

File: src/core/ext/filters/client_channel/backup_poller.cc
Comment (apolcyn) [reply src/core/ext/filters/client_channel/backup_poller.cc]: removed

File: src/core/ext/filters/client_channel/health/health_check_client.cc
Comment (apolcyn) [reply src/core/ext/filters/client_channel/health/health_check_client.cc]: removed

File: src/core/ext/filters/client_channel/lb_policy/priority/priority.cc
Comment (apolcyn) [reply src/core/ext/filters/client_channel/lb_policy/priority/priority.cc]: removed

File: src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc
Comment (apolcyn) [reply src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc]: removed

File: src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_manager.cc
Comment (apolcyn) [reply src/core/ext/filters/client_channel/lb_policy/xds/xds_cluster_manager.cc]: removed

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc
Comment (apolcyn) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc]: The overall resolution timeout isn't repeated, but the backup poller alarm initiated just after is. I've removed the timer invalidation to be just above the spot where we initialize the backup polling loop

File: src/core/ext/filters/client_channel/subchannel.cc
Comment (apolcyn) [reply src/core/ext/filters/client_channel/subchannel.cc]: removed

File: src/core/ext/filters/client_channel/retry_filter.cc
Comment (apolcyn) [reply src/core/ext/filters/client_channel/retry_filter.cc]: removed

File: src/core/ext/filters/fault_injection/fault_injection_filter.cc
Comment (apolcyn) [reply src/core/ext/filters/fault_injection/fault_injection_filter.cc]: removed

**Comment by markdroth** [comment]
The code changes LGTM, modulo the chttp2 change, which @yashykt should review.

Can you please add a comment above the `InvalidateNow()` calls you're adding to indicate why they're needed?  Otherwise, future readers may remove them, not understanding why they're needed.

Also, is there a way to add tests for these specific cases?  I don't want to block this fix on that if the tests are too hard to write, but if they aren't too hard, it seems better to have them.

**Comment by yashykt** [comment]
I'll review in a bit. I haven't yet read through the issue yet.

File: src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc
Comment (apolcyn) [reply src/core/ext/filters/client_channel/resolver/dns/c_ares/grpc_ares_wrapper.cc]: Sorry on second though, I ended up going back and removing this, since even though this `grpc_timer_init` initiates the backup poller, the backup poller uses it's own callback and it's own `grpc_timer_init`, so this `grpc_timer_init` indeed isn't repeated.

File: src/core/ext/transport/chttp2/transport/writing.cc
Comment (yashykt) [reply src/core/ext/transport/chttp2/transport/writing.cc:81]: This does run within a combiner if that's what the original question is

Comment (yashykt) [reply src/core/ext/transport/chttp2/transport/writing.cc:81]: The more I think about this, the more I'm thinking that we should either not cache `Now()` at all or maybe we should add `InvalidateNow()` at some more general locations, for example, when closures are executed on the work serializer/combiner

**Review by yashykt — APPROVED** [review]
LGTM with the understanding that this PR should be reverted once we decide on the right way to fix this

**Comment by apolcyn** [comment]
I agree it would be great to get a unit test for this, but it indeed appears difficult to write such a test, and a benchmark style test like the one mentioned in https://buganizer.corp.google.com/issues/185205431#comment15 is IMO the most effective way to test for this, since the failure mode manifests itself with some operations taking much longer than they should (but they can still complete). Also, the problem seems to repro so far only on single-core VMs (which we can be easily controlled for in an e2e benchmark style test).

(after this PR is imported, I plan to run the repro from https://buganizer.corp.google.com/issues/185205431#comment15 continuously).

**Comment by apolcyn** [comment]
basic tests C/C++ macos: https://github.com/grpc/grpc/issues/26110

Bazel C/C++ opt macos looks like an issue with bazel/RBE itself:

```
+ tools/bazel --bazelrc=tools/remote_build/mac.bazelrc test --invocation_id=4937b338-36f4-423c-99f0-8eaef241479e --workspace_status_command=tools/remote_build/workspace_status_kokoro.sh --google_credentials=/tmpfs/src/gfile/GrpcTesting-d0eeee2db331.json --config=opt -- //test/...
INFO: Running bazel wrapper (see //tools/bazel for details), bazel version 3.7.1 will be used instead of system-wide bazel installation.
WARNING: Running Bazel server needs to be killed, because the startup options are different.
Starting local Bazel server and connecting to it...
INFO: Invocation ID: 4937b338-36f4-423c-99f0-8eaef241479e
ERROR: java.io.IOException: Error getting access token for service account: 400 Bad Request
```

**Review by markdroth — APPROVED** [review]

**Pull request merged**
